Python/Product/Analysis/ProjectEntry.cs

```diff
@@ -342,15 +316,101 @@ where fileName.StartsWith(pathPrefix)
             }
         }
 
-        #endregion
-
         internal void Enqueue() {
             _unit.Enqueue();
         }
 
         public void AggregatedInto(AggregateProjectEntry into) {
             _aggregates.Add(into);
         }
+
+        public IEnumerable<int> DocumentParts => _buffers.Keys.AsLockedEnumerable(_buffers);
```

Comment:
This is pretty much the same as `lock(this)`. Any client code that calls `DocumentParts.GetEnumerator()` acquires a lock that controls access to private implementation.

Reply:
When used correctly (which is to enumerate this sequence and use the values to index into `_buffers`) it's going to be necessary to lock the outer object on all uses. This way, we get a single lock that lasts for the entire `foreach` and prevents races.

Either way requires the caller to use the object correctly. Not disposing an `IDisposable` is more likely to be done correctly than locking the object everywhere.